c_src/main.c

```diff
@@ -78,17 +78,17 @@ erlfdb_future_cb(FDBFuture* fdb_future, void* data)
     ERL_NIF_TERM msg;
     bool cancelled;
 
-
-    enif_mutex_lock(future->lock);
+    enif_mutex_lock(future->cancel_lock);
     cancelled = future->cancelled;
-    enif_mutex_unlock(future->lock);
+    enif_mutex_unlock(future->cancel_lock);
 
     if(!cancelled) {
+        enif_mutex_lock(future->msg_lock);
         msg = T2(future->msg_env, future->msg_ref, ATOM_ready);
         enif_send(NULL, &(future->pid), future->msg_env, msg);
+        enif_mutex_unlock(future->msg_lock);
     }
```

Comment:
@jhogberg really appreciated for your comments!

We are stuck here, for what to set for the `caller_env` in `enif_send`. 
It is clear that we could use `NULL` from callback in a 3rd party managed thread.

However as you mentioned in https://github.com/erlang/otp/issues/8358#issuecomment-2042961991, it is tricky to handle the sync callback from the 3rd party which is executed by beam managed scheduler thread.

In the PR https://github.com/emqx/couchdb-erlfdb/pull/18/files#diff-ceadfa165520612e5b6b1a255f7680bcaa9e2a67bf47677f6884fd1db3cb03f5L122 it tried to fix the issue of reusing the `env` of previous nif function call, as the `env` becomes invalid after that NIF function call. The old behaviour is to save the caller's env and reuse it in the caller.  So we change it to `NULL` to pretend the `enif_send` is called from a none beam managed and it seems to work in our CI env. 

as I see you propose following: 

> thread-specific data is probably the most straightforward way

Do you mean we could use "process independent environments " ? We are a bit unsure about that because it was not stated in the doc about `call_env`
```
caller_env
The environment of the calling thread ([process bound](https://www.erlang.org/doc/man/erl_nif#proc_bound_env) or [callback](https://www.erlang.org/doc/man/erl_nif#callback_env) environment) or NULL if calling from a custom thread not spawned by ERTS.
```

BTW I look into the `enif_send` implementation and I couldn't see set `caller_env` to NULL in scheduler thread could cause issue as when `caller_env` is set to NULL, the `c_p` is set to NULL then the side effect of it is just send the `from` to undefined.  I am double checking, will be appreciated if you could give me some hints.

Reply:
Is the process-bound env safe to pass here? what if the process is dead or moved to other scheduler thread?